armsrc/em4x50.c

```diff
@@ -0,0 +1,877 @@
+    
+    // returns bit value (or EM4X50_BIT_OTHER -> no bit pattern) by evaluating
+    // a single sample within a bit period (given there is no LIW, ACK or NAK)
+    // This function is not used for decoding, it is only used for identifying
+    // a listen window (return value = EM4X50_BIT_OTHER) in functions
+    // "find_double_listen_window" and "check_ack"
+  
+    int dhigh = 230;    // 90% of maximum sample value (255)
+    int dlow = 25;      // 10% of maximum sample value (255)
+    uint8_t sample;
+
+    // get sample at 3/4 of bit period
+    wait_timer(0, T0 * EM4X50_T_TAG_THREE_QUARTER_PERIOD);
+    sample = (uint8_t)AT91C_BASE_SSC->SSC_RHR;
+
+    // wait until end of bit period
+    wait_timer(0, T0 * EM4X50_T_TAG_QUARTER_PERIOD);
+
+    // decide wether "0" or "1"
+    if (sample > dhigh)
+        return EM4X50_BIT_0;
+    else if (sample < dlow)
+        return EM4X50_BIT_1;
+    
+    return EM4X50_BIT_OTHER;
+}
+
+static uint32_t get_pulse_length(void) {
+    
+    // iterates pulse length (low -> high -> low)
+    
+    uint8_t sample = 0, high = 192, low = 64;
+
+    sample = (uint8_t)AT91C_BASE_SSC->SSC_RHR;
+    
+    while (sample > low)
+        sample = (uint8_t)AT91C_BASE_SSC->SSC_RHR;
+
+    AT91C_BASE_TC1->TC_CCR = AT91C_TC_SWTRG;
+
+    while (sample < high)
+        sample = (uint8_t)AT91C_BASE_SSC->SSC_RHR;
+
+    while (sample > low)
+        sample = (uint8_t)AT91C_BASE_SSC->SSC_RHR;
+
+    return (uint32_t)AT91C_BASE_TC1->TC_CV;
+}
+
+static bool check_pulse_length(uint32_t pl, int length) {
+    
+    // check if pulse length <pl> corresponds to given length <length>
+    
+    if ((pl >= T0 * (length - EM4X50_TAG_TOLERANCE)) &
+        (pl <= T0 * (length + EM4X50_TAG_TOLERANCE)))
+        return true;
+    else
+        return false;
+}
+
+static void em4x50_send_bit(int bit) {
+    
+    // send single bit according to EM4x50 application note and datasheet
+    
+    // reset clock for the next bit
+    AT91C_BASE_TC0->TC_CCR = AT91C_TC_SWTRG;
+
+    if (bit == 0) {
+
+        // disable modulation (drop the field) for 7 cycles of carrier
+        // period (Opt64)
+        LOW(GPIO_SSC_DOUT);
+        while (AT91C_BASE_TC0->TC_CV < T0 * 7);
+
+        // enable modulation (activates the field) for remaining first
+        // half of bit period
+        HIGH(GPIO_SSC_DOUT);
+        while (AT91C_BASE_TC0->TC_CV < T0 * EM4X50_T_TAG_HALF_PERIOD);
+        
+        // disable modulation for second half of bit period
+        LOW(GPIO_SSC_DOUT);
+        while (AT91C_BASE_TC0->TC_CV < T0 * EM4X50_T_TAG_FULL_PERIOD);
+
+    } else {
+        
+        // bit = "1" means disable modulation for full bit period
+        LOW(GPIO_SSC_DOUT);
+        while (AT91C_BASE_TC0->TC_CV < T0 * EM4X50_T_TAG_FULL_PERIOD);
+    }
+}
+
+static void em4x50_send_byte(uint8_t byte) {
+
+    // send byte (without parity)
+    
+    int mask = 0x80;
+
+    for (int i = 0; i < 8; i++) {
+
+        if ((byte & mask) == 0)
```

Comment:
I think you can simplify this if statement to one line...

Reply:
done